<!-- -*- mode: XML -*- -->
<!-- run velodyne_pointcloud/CloudNodelet in a nodelet manager for a VLP-16 -->

<launch>

  <group ns="M_HDL32">
    <!-- declare arguments with default values -->
    <arg name="calibration" default="$(find velodyne_pointcloud)/params/32db.yaml"/>
    <arg name="device_ip" default="192.168.140.201" />
    <arg name="frame_id" default="M_velodyne" />
    <arg name="manager" default="$(arg frame_id)_nodelet_manager" />
    <arg name="max_range" default="130.0" />
    <arg name="min_range" default="0.4" />
    <arg name="pcap" default="" />
    <arg name="port" default="2368" />
    <arg name="read_fast" default="false" />
    <arg name="read_once" default="false" />
    <arg name="repeat_delay" default="0.0" />
    <arg name="rpm" default="600.0" />
    <arg name="laserscan_ring" default="-1" />
    <arg name="laserscan_resolution" default="0.007" />

    <!-- start nodelet manager and driver nodelets -->
    <include file="$(find velodyne_driver)/launch/nodelet_manager.launch">
      <arg name="device_ip" value="$(arg device_ip)"/>
      <arg name="frame_id" value="$(arg frame_id)"/>
      <arg name="manager" value="$(arg manager)" />
      <arg name="model" value="32E"/>
      <arg name="pcap" value="$(arg pcap)"/>
      <arg name="port" value="$(arg port)"/>
      <arg name="read_fast" value="$(arg read_fast)"/>
      <arg name="read_once" value="$(arg read_once)"/>
      <arg name="repeat_delay" value="$(arg repeat_delay)"/>
      <arg name="rpm" value="$(arg rpm)"/>
    </include>

    <!-- start cloud nodelet -->
    <include file="$(find velodyne_pointcloud)/launch/cloud_nodelet.launch">
      <arg name="calibration" value="$(arg calibration)"/>
      <arg name="manager" value="$(arg manager)" />
      <arg name="max_range" value="$(arg max_range)"/>
      <arg name="min_range" value="$(arg min_range)"/>
    </include>
  </group>

  <group ns="FL_VLP16">
    <!-- declare arguments with default values -->
    <arg name="calibration" default="$(find velodyne_pointcloud)/params/VLP16db.yaml"/>
    <arg name="device_ip" default="192.168.140.202" />
    <arg name="frame_id" default="FL_velodyne" />
    <arg name="manager" default="$(arg frame_id)_nodelet_manager" />
    <arg name="max_range" default="130.0" />
    <arg name="min_range" default="0.4" />
    <arg name="pcap" default="" />
    <arg name="port" default="2369" />
    <arg name="read_fast" default="false" />
    <arg name="read_once" default="false" />
    <arg name="repeat_delay" default="0.0" />
    <arg name="rpm" default="600.0" />
    <arg name="laserscan_ring" default="-1" />
    <arg name="laserscan_resolution" default="0.007" />

    <!-- start nodelet manager and driver nodelets -->
    <include file="$(find velodyne_driver)/launch/nodelet_manager.launch">
      <arg name="device_ip" value="$(arg device_ip)"/>
      <arg name="frame_id" value="$(arg frame_id)"/>
      <arg name="manager" value="$(arg manager)" />
      <arg name="model" value="VLP16"/>
      <arg name="pcap" value="$(arg pcap)"/>
      <arg name="port" value="$(arg port)"/>
      <arg name="read_fast" value="$(arg read_fast)"/>
      <arg name="read_once" value="$(arg read_once)"/>
      <arg name="repeat_delay" value="$(arg repeat_delay)"/>
      <arg name="rpm" value="$(arg rpm)"/>
    </include>

    <!-- start cloud nodelet -->
    <include file="$(find velodyne_pointcloud)/launch/cloud_nodelet.launch">
      <arg name="calibration" value="$(arg calibration)"/>
      <arg name="manager" value="$(arg manager)" />
      <arg name="max_range" value="$(arg max_range)"/>
      <arg name="min_range" value="$(arg min_range)"/>
    </include>
  </group>

  <group ns="FR_VLP16">
    <!-- declare arguments with default values -->
    <arg name="calibration" default="$(find velodyne_pointcloud)/params/VLP16db.yaml"/>
    <arg name="device_ip" default="192.168.140.203" />
    <arg name="frame_id" default="FR_velodyne" />
    <arg name="manager" default="$(arg frame_id)_nodelet_manager" />
    <arg name="max_range" default="130.0" />
    <arg name="min_range" default="0.4" />
    <arg name="pcap" default="" />
    <arg name="port" default="2370" />
    <arg name="read_fast" default="false" />
    <arg name="read_once" default="false" />
    <arg name="repeat_delay" default="0.0" />
    <arg name="rpm" default="600.0" />
    <arg name="laserscan_ring" default="-1" />
    <arg name="laserscan_resolution" default="0.007" />

    <!-- start nodelet manager and driver nodelets -->
    <include file="$(find velodyne_driver)/launch/nodelet_manager.launch">
      <arg name="device_ip" value="$(arg device_ip)"/>
      <arg name="frame_id" value="$(arg frame_id)"/>
      <arg name="manager" value="$(arg manager)" />
      <arg name="model" value="VLP16"/>
      <arg name="pcap" value="$(arg pcap)"/>
      <arg name="port" value="$(arg port)"/>
      <arg name="read_fast" value="$(arg read_fast)"/>
      <arg name="read_once" value="$(arg read_once)"/>
      <arg name="repeat_delay" value="$(arg repeat_delay)"/>
      <arg name="rpm" value="$(arg rpm)"/>
    </include>

    <!-- start cloud nodelet -->
    <include file="$(find velodyne_pointcloud)/launch/cloud_nodelet.launch">
      <arg name="calibration" value="$(arg calibration)"/>
      <arg name="manager" value="$(arg manager)" />
      <arg name="max_range" value="$(arg max_range)"/>
      <arg name="min_range" value="$(arg min_range)"/>
    </include>
  </group>

  <rosparam command="load" ns="velodyne_calibration" file="$(find safe_traffic_weaving)/scripts/multi_velodyne_calibration.yaml" />

  <node pkg="tf" type="static_transform_publisher" name="M_velodyne_tf_broadcaster" args="0 0 1.47 0 0 -0.70710678 0.70710678 vehicle_base M_velodyne 100" />
  <node pkg="tf" type="static_transform_publisher" name="FL_velodyne_tf_rough_broadcaster" args="2.13  1.02 0.66 0 0 0 1 vehicle_base FL_velodyne_rough 100" />
  <node pkg="tf" type="static_transform_publisher" name="FR_velodyne_tf_rough_broadcaster" args="2.13 -1.02 0.66 0 0 0 1 vehicle_base FR_velodyne_rough 100" />
  <node pkg="tf" type="static_transform_publisher" name="FL_velodyne_tf_broadcaster" launch-prefix="/bin/bash -c '$* `rosparam get velodyne_calibration/FL_VLP16_args`' --" />
  <node pkg="tf" type="static_transform_publisher" name="FR_velodyne_tf_broadcaster" launch-prefix="/bin/bash -c '$* `rosparam get velodyne_calibration/FR_VLP16_args`' --" />
</launch>
